src/deluge/model/clip/instrument_clip.cpp

```diff
@@ -1495,9 +1491,12 @@ int32_t InstrumentClip::getYNoteFromYVisual(int32_t yVisual, Song* song) {
 int32_t InstrumentClip::guessRootNote(Song* song, int32_t previousRoot) {
 	NoteSet notesPresent;
 
-	seeWhatNotesWithinOctaveArePresent(
-	    notesPresent, 0, song,
-	    false); // Don't delete anything yet, since we're still going to make use of the noteRowsOnScreen!
+	seeWhatNotesWithinOctaveArePresent(notesPresent, song->key);
+	// TODO: there used to be "deleteUnusedRows" argument, and this passed in false, but that
+	// didn't actually do anything. Check that it wasn't just a bug, and not deleting would make
+	// things right.
+	//
```

Comment:
The two users I can see are both in `Song` -- one in `Song::setRootNote` and the other in `Song::setPresetScale`. However, in the implementation, that argument is unused. I don't think this is a bug strictly speaking, but it *is* a potential place to save memory and gain some performance. If we delete the now-empty `NoteRows` those won't be processed for sequencing and we can reuse that memory for something else later. It's probably wise to actually implement that in another PR if we're going to do it.

Reply:
We already do delete the empty note rows, unconditionally. This wanting to keep them is what led me to wonder.